### ISSUE TITLE
Add opt-in mechanisms for a shortcut to RE(...) syntax

### ISSUE THREAD

**Issue by danielballan** [bluesky/bluesky#1238]
From recent discussions with @prjemian, @tacaswell, @klauer and months-old discussions with @teddyrendahl and @ZLLentz:

After years of due caution, it seems we might be ready to pull the trigger on giving The People what they have been asking for for a long time: some ways to skip typing "RE" to execute plans. There are three proposals for how to do this. I propose to add all three of these to the codebase as *opt-in mechanisms* and mark them as "experimental" and subject to future removal if they prove problematic of confusing in practice. I do not think we have consensus on which one is best, so a good way forward may be to present them all to users and learn from experience.

### Proposed behavior

If the user types a bluesky plan like `count([det])` as the last (or only) line in an IPython input, assume that they want to execute it. This does not prevent them from doing other things with plans; typing `summarize_plan(count([det]))` would still work the same as now.

### Proposed implementation

To opt in, the user, typically in startup configuration do something like:

```python
bluesky.utils.enable_auto_exec(RE)
```

where a specific `RE` instance must be passed in.

All built-in plans (including stubs) would be wrapped in a new `@bluesky.plan` decorator. This decorator gives us a way to recognize that an object is a plan (as opposed to some other generator instance or iterable).

The decorator has some side benefits as well:
* The `@plan` decorator would mark nicely in the code that something is a plan.
* The decorator could give the plan a nice `__repr__` so that, when "auto exec" mode is _disabled_, the user who types a plan and expects to run would see a clear message that this is a plan. (This has been proposed previously but not yet done. There are also other ways to accomplish it.)

### Future possiblities

There has been early discussion of adding methods to some ophyd Device that return plans, to express multi-step sequencing that is best stored on a Device. We could use this mechanism to support usages like:

```python
shutter.close()
```

where `shutter.close()` returns a plan but typing it at the REPL in IPython causes the plan to be run. Thus, we do things "the bluesky way", async and relying on the RunEngine to manage execution, but still support a fluid interactive usage.

### Proposed behavior

The user types some magic like `%RE_mode on` and then _every_ future user input is passed to `RE(...)`. If any input is typed that is not a plan, such as `db[-1]`, it will be passed to `RE` and whatever error will be raised.  To do `db[-1]` the user would type `%RE_mode off`. When the mode is on, the PS1 will be modified to make clear what mode we are in.

### Proposed implementation

Similarly to Mechanism 1, to set up RE mode, the user, typically in startup configuration do something like:

```python
bluesky.utils.configure_RE_mode(RE)
```

where a specific `RE` instance must be passed in.

## Mechanism 3: Special character input transformer

If we user types

```
> count([det])
```

we display

```
--> RE(count([det]))
```
and the run that. This has been in use at HXN, and they are happy with it. It actually already in bluesky, so the action item here would be to document it and let it be known more widely.

# Next step

Does anyone feel strongly enough that one of these is a _bad_ idea to block it going in?

**Comment by prjemian** [comment]
*Mechanism 2: a modal plan-only mode* is not bad but is the weakest of these three.  Too easy for users to forget which mode and thus introduce unexpected behavior.  Most likely (of these three) to cause problems.

You did not (yet) ask us to vote up any.

**Comment by danielballan** [comment]
I agree with that. Specifically, I notice that Mechanisms 1 and 3 are _compatible_ with the normal `RE(...)` usage, so if users follow the documentation, what the docs say will still work. With Mechanism 2, if I'm in "RE mode", the normal usage would fail.

**Comment by klauer** [comment]
(1) seems like the best option I've heard of yet. Continuing to maintain (3) seems to be not very costly. 

I'm fully on board with moving forward with at least (1) and (3) though I'm not so crazy about (2).

**Comment by teddyrendahl** [comment]
Fine with having them all be in the codebase and opt-in if you have that bandwidth. Otherwise, if you are picking just one I'm by far the biggest fan of (1)

**Comment by prjemian** [comment]
(1) for sure.  It could enable fulfillment of a request to list all known
plans.

**Comment by awalter-bnl** [comment]
Just a a quick point, this is a detail but just wanted to capture it.

yesterday one of our beamline staff ran into an issue which he didn't understand at NSLS-II which relates to the first option, and in particular this line:
>All built-in plans (including stubs) would be wrapped in a new @bluesky.plan decorator.

Some plan stubs should probably not be wrapped in the decorator:
For instance the plan stub `trigger_and_read` can not be fed into the run engine by itself, as it contains both a `yield from create()` and a `yield from save()`, but does not also contain the pre-requisite `yield from open_run()` and `yield from close_run()` that these require and hence raises an error if used in the run_engine by itself. Therefore doing `RE(trigger_and_read(...))` raises an error (Note: the beamline scientist attempted yesterday to run a `per_step` partial plan directly in the `run engine` and ran into trouble because it contained a `yield from trigger_and_read(...)` in it !-) ) .

Therefore `trigger_and_read`, and other plan_stubs like it, should probably not be decorated. Others like `mv` and `mvr` definately should IMO however. So I guess the line quoted above maybe should be modified to:
>All built-in `plans` (and `plan stubs` that can be passed directly into the `run engine`) would be wrapped in a new `@bluesky.plan` decorator.

**Comment by danielballan** [comment]
That is worth some thought. But wouldn’t there still be confusion? The scientist would have type trigger_and_read(...) and wondered, “Why won’t this plan run?!” If they haven’t yet encountered the distinction between plans that can be run as is and plans that require some assembly to use, wouldn’t they still end up confused and stuck? Would a better solution be to improve our error message?

**Comment by prjemian** [comment]
Different decorator for subs that need help to run as plan?

**Comment by danielballan** [comment]
Or an option to the decorator:

```python
@plan(if_run_alone_raise=IllegalMessageSequence("This needs to be used inside a plan that..."))
def trigger_and_read(...):
    ...
```

But this approach would _only_ help users that were operating in IPython with one of these "mechanisms" on. I'd prefer a solution that helps users who are using the vanilla `RE(trigger_and_read())` usage too. Let's move this to #1240 to avoid side-tracking this issue. Thanks for raising the point, @awalter-bnl.

**Comment by awalter-bnl** [comment]
> Would a better solution be to improve our error message?

Agree this would probably be a better solution.

**Comment by danielballan** [comment]
I solicited some feedback from Stuart Wilkins as well, who weighed in in favor of Mechanism 1 as well. Via Slack:

> FWIW I really like [Mechanism] 1. Seems very pythonic.

**Comment by teddyrendahl** [comment]
> Some plan stubs should probably not be wrapped in the decorator
My initial impression was that none of the stubs would be wrapped, and only fully complete plans would get the wrapper. This in a way would tell a user very clearly if the plan was intended to run on its own or not.

I also would add, it would be nice if the decorator handled some default metadata. I feel like the first 12 lines of any plan is the same boiler plate handling of user input metadata, plus my defaults. I think this wouldn't be so hard to include?

**Comment by awalter-bnl** [comment]
@teddyrendahl, what about the `plan_stubs` `mv` and `mvr` for instance. I think them (and other similar `plan_stubs`) should be wrapped. These are `plan_stubs` that _can_ be run on their own.

I do agree that having the decorator handle some of the default meta-data may be useful, but if we expect custom defined plans to also be able to run the decorator then what can/can't be included in the decorator becomes difficult to know.

**Comment by danielballan** [comment]
Agreed with @awalter-bnl  about `mv[r]`. In fact, the majority of plan stubs make sense on their own in a debugging context: `trigger`, `read`, etc. Only the state-management plans (create, save, drop) stand out. We'll have to decide if/how to mark them as special. I know @tacaswell is preparing to weigh in against Mechanism (1) altogether...maybe we should wait for that before we dig into details too deeply.

> I also would add, it would be nice if the decorator handled some default metadata. I feel like the first 12 lines of any plan is the same boiler plate handling of user input metadata, plus my defaults. I think this wouldn't be so hard to include?

Interesting. It would be nice to streaming that metadata handling. It might help encourage users to follow the best practices modeled by `bluesky/plans.py`. I might argue to tackling this in a second pass, but it sounds both worthwhile and doable.

**Comment by tacaswell** [comment]
A third advantage of the decorator is that we could have a `__del__` method to warn when it is gc'd without being yielded from.  One drawback of the the decorator is that our tracebacks are going to get even more convoluted.


-----

I think that maintaining a difference between "code that hardwares" and "code that maths" is important.  Consider seeing the line

```
In[100]: xpcs()
```

in a beamline terminal or when scrolling up through history.  Did that mean:

```python
def xpcs(scan_id=-1, **kwargs):
    h = db[scan_id]
    return plot_g2_given_header(h, **kwargs)
```

or

```python
def xpcs(num_frames=10000, exp_time=.1):
    for d in dets:
         yield from bps.mov(d.exposure_time, exp_time)
    return (yield from the_actual_plan(dets, num_frames,))
```

Both of those are reasonable functions that I would not be surprised to find in beamline profiles.  If we are using some sort of ceremony (e.g. `RE(...)` or `< ...` to visually indicate "this one does hardware" it, removes that ambiguity.  To pull out my least favorite trope, at 3AM at the beamline you want the user to be very sure they are about to start 6 hours of data collection, not 6 hours of computation.

As others have pointed out, there is some ambiguity as to _which_ plans should auto run.  This could cut both ways where either the user expects something to auto run (and it does not) or they expect it is "low level" enough to not run and then it does which will lead to some combination of frustration and unexpected hardware operation.

This ambiguity is also setting traps for intermediate users.  Everyone chiming in here understands plans and how the RE works so you _know_ that when you write your own plans by composing existing plans you have to use `yield from`, but for users that are just starting to write their own plans is not clear.   At some fundamental level, plans are not "just a python function call" as we rely on the RE to manage all sorts of things and trying to elide this _sometimes_ is worse than the current status-quo.  Users are going to have to remember which functions are plans ("do hardware" and which are not ("do math") no matter what, adding a sometimes magic requires them to also remember which ones have it and in which contexts it applies.

The biggest trap may be if 


```python
In [1]: count()
```

works, then it is super weird that:

```python
In [2]: for j in range(5): 
   ...:     count() 
   ...:     
```

does not.

-----

For the modal option, I think we can make the PS1 change drastic enough to make it very clear, maybe changing it to `RE: ` or some such, maybe change the color of the prompt?  I am optimistic we can make this very obvious.  Given the number of vi users we have around I am surprised that this is not the favorite :smiling_imp: .

----

option 3 is far and away my favorite of these.  It is the tersest possible way to provide some ceremony to differentiate between "code that will do hardware" and "code that will do math".  You get 99% of the benefit of option 1 (at the cost of one extra character) with none of the down sides.

----

A combination of 2 and 3 where there are escapes to the other mode seems interesting.

----


There is an option 4 which is having the decorator walk the AST of the function it is wrapping and adding `yield from`s that it think it needs.  I am very much against this approach (as it is essentially creating a python-like DSL), but option 1 feels like it is trying to do a "lite" version of this "on the cheap".  On the other hand, given my comments above about holding a distinction between "do hardware" and "do math", maybe we want to seriously consider this?


----

Another benefit of the decorator is that this is a step in the direction of implementing a scan algera so 

```python
RE((count(...) + bps.mv(..) + count() + bps.mv(...)) * 5)
```

is a thing that works like you expect :imp:

**Comment by danielballan** [comment]
Thanks, @tacaswell. A `@plan` decorator is starting to sound like a good idea _regardless_ of whether we bless people using it to auto-run plans. Let's move the discussion about the decorator's design and tradeoffs to #1239.

**My goal in raising this issue is to stem the split into plans and "convenience function" that call `RE`.** In the wild, many beamlines are making functions that call `RE`, such as `snap()` and `lights.off()` to name two from BMM. In four years, we have failed to persuade users not to do this, even users who generally praise bluesky and strive to work with it rather than against it. I conclude that, to heal this split, we have to give users something close enough to what they want.

I also foresee a coming conflict between whether we settle on:

```py
# This is "obviously right" to people who primarily use ophyd interactively.
device.stage().plan()  # returns a plan
device.stage()  # executes
```

or

```py
# This is "obviously right" to people who primarily use ophyd inside plans.
device.stage()  # returns a plan
device.stage().go()  # executes
```

and I think this may let us avert that conflict.

This point about `for` loops is a good one:

> The biggest trap may be if
```py
In [1]: count()
```
> works, then it is super weird that:

```py
In [2]: for j in range(5): 
   ...:     count() 
   ...:     
```
> does not.

But isn't that true for all of these proposals? A `for` loop wouldn't work in Mechanisms (2) or (3) either. Based on feedback so far, I think using the decorator to guess the right mode is the mechanism that will persuade folks to stop writing convenience functions. Like we currently do with `>` at HXN, we could print what is being done:

```py
In [3]: count([det])
--> RE(count([det]))
```

I remain open to providing all three and getting wider feedback from users on all of them if no one has strong objections.

**Comment by tacaswell** [comment]
> A `for` loop wouldn't work in Mechanisms (2) or (3) either.

but you also would not _expect_ it to.

> In four years, we have failed to persuade users not to do this,

We also have not aggressively documented or advertised option 3 and have not written option 1.  This is a case that we need to do _something_, not really an argument for any of them in particular.

> I also foresee a coming conflict between whether we settle on:

Again, option 3 also avoids this conflict without any of the downsides of option 1 and way less magic.

**Comment by danielballan** [comment]
Well, let's get a temperature check on this. Before @tacaswell's comment, it sounded like his favorite is everyone else's _least_ favorite and vice versa. How do things stand in light of the subsequent discussion, @teddyrendahl @prjemian @klauer @awalter-bnl (and anyone else who hasn't chimed in yet)?

**Comment by prjemian** [comment]
On 2019-08-14 1:27 PM, Dan Allan wrote:
> In the wild, many beamlines are making functions that call |RE|

Seen while on safari at APS, as well.  Convenience vs. elegance.
Some still use tcsh although usage of punch cards is quite low.

**Comment by awalter-bnl** [comment]
I am up for putting all of them out temporarily and seeing which ones users prefer. 

I _suspect_ option 2 will be widely ignored and that option 1 will be most widely adopted but I see all of these as improvements on what is currently already being done.

**Comment by prjemian** [comment]
* Definite downvote on 2 (modal plan-only mode) - there will be unintended behavior whether or not PS1 is changed - people are blind
* OK using some character (such as `< bp.scan(...)`) to replace `RE(bp.scan(...))`; it **is** a positive step.  Fewer parentheses. This is method 3.
* I really like the idea of a `@bp.plan` decorator.  It is an aid in user-written files that mix *math* and *hardware* functions.  It's the hardware functions that have the different behavior to be learned for new users.

**Comment by danielballan** [comment]
I propose these initial steps. :+1:s?

- Create the `@plan` decorator for all the other benefits discussed (assuming the stack depth increase doesn't turn out to be awful).
- Document and start advertising Mechanism (3), the `<`.
- Revisit adding Mechanism (1) once the decorator is merged (but maybe we decide not to do it).
- Forget about Mechanism (2) (modal) for now because it is no one's favorite and actively disliked by some.

**Comment by ZLLentz** [comment]
Chiming in a little late:

I used to be a fan of mechanisms like these. However, the more I think about `bluesky` the more I like the explicit separation between "calculating" and "doing". In that regard, I agree that mechanism 3 is the best option of these three and I hope people enjoy using it, and as it has already been implemented it's clearly the way to start. It is terse and frees us from the easily missed and visually confusing double parentheses. My only objection is that it feels like an unnatural extension to Python (which is an argument you should ignore).

**Comment by bruceravel** [comment]
Chiming in extremely late ... I was lead here by @danielballan mentioning me in https://github.com/bluesky/bluesky-enhancement-proposals/issues/18

From my perspective as a BlueSky consumer and beamline scientist, I dislike both Mechanism 2 and 3.  The common reason is that neither one will make it easier for me to teach my users how to make an experiment happen.

My problem with mechanism 2 is that the user mostly types things that are plans, but occasionally types things that are not plans.  With modal input, I have to train the users on the difference between the two.  Even so, I am likely to get that 10:00 pm phone call asking about what went wrong because they forgot about switching back into RE mode.

My problem with mechanism 3 is that is does exactly nothing to lower the bar in learning how to interact with the beamline, it just changes the spelling.  It is, I suppose, easier to type `> ...`  than `RE(...)`.  But it does not make it any easier to explain what needs to be done and why it needs to be done that way.

My preference is mechanism 1 as explained by Dan and with his "future possibility".  This puts the burden of doing the right thing on me, the fellow developing the beamline code.  It flattens the space of ways that the user interacts with the beamline.  That is, `shutter.open()` and a quick calculation (i.e. `sqrt(643/3.81)` to find out what photoelectron wavenumber approximately corresponds to 643 eV above the absorption edge, a sort of calculator interaction that happens daily at my BL) are both just "things a user types."  That seems great to me.

A while back, @tacaswell helped me with another partial solution to this problem: https://github.com/NSLS-II-BMM/profile_collection/blob/master/startup/94-prompt.py#L67-L68 When the user types, say, `count(dets)`, ipython responds with  something like:
```
Out[204]: <generator object count at 0x7fc9a8c19750>   Hint: enclose that command in RE()
```
Given that I currently have to explain what `RE(...)` is and why it must be typed every time, that response is genuinely useful and **will** save me a call at 10:00 pm.

**Comment by cjtitus** [comment]
I would like to wake this issue up, dust it off, and see what people think about it. Quite a lot of time has passed since it was originally discussed. I am determined to drag plan decorators across the finish line, and this is a logical thing to think about once that is done. I would like to get ahead of the curve, especially in the case that ideas here would impact the implementation of plan decorators.

Have people's opinions changed? Does anybody have additional concerns, ideas, improvements to the concept?
